Add and Edit configmap form

karthikjeeyar opened the pull request:
**Fixes**: 
<!-- For e.g Fixes: https://issues.redhat.com/browse/ODC-XXX -->
https://issues.redhat.com/browse/ODC-6483
https://issues.redhat.com/browse/ODC-6484


**Analysis / Root cause**: 
<!-- Briefly describe analysis of US/Task or root cause of Defect -->
Add Create and Edit configmap form/yaml switcher

**Solution Description**: 
<!-- Describe your code changes in detail and explain the solution -->
Added Form/Yaml based create and edit forms for configmap.

**Screen shots / Gifs for design review**: 
<!-- If change affects UI in any way, tag @openshift/team-devconsole-ux and add screenshots/gifs  -->
**Create configmap:**

![image](https://user-images.githubusercontent.com/9964343/160435011-eb68ca88-dce2-4e7f-b6b2-45a28185995b.png)

Video:

https://user-images.githubusercontent.com/9964343/159980191-3cf91e1a-285e-4c2a-9dfe-2d06d2fb0e78.mov

**Edit Configmap:**

![image](https://user-images.githubusercontent.com/9964343/160435120-1af5f38a-146c-4f7f-ac7d-7fec821b8f13.png)



Video:

https://user-images.githubusercontent.com/9964343/159980211-58e2cf01-d72b-46fa-b4d4-384f8eef454a.mov


**Unit test coverage report**: 
<!-- Attach test coverage report -->

KeyValueFileInputField.spec.tsx
```
  ✓ should have validation error given input field is touched and error exists on form (107ms)
  ✓ should have remove key value pair button if there are more than one key value entries (36ms)
  ✓ should not contain remove key value pair button if there is one entry (14ms)
  ✓ should contain remove key value pair button if there is one entry and remove actions is set to false (15ms)
  ✓ should add new entry on clicking Add key/value button (41ms)
```

  ```
   configmap-utils
    isBase64:
      ✓ should return false for invalid inputs (1ms)
      ✓ should return true for valid inputs
    getInitialConfigMapFormData
      ✓ should return initialFormData if configMap is not passed (1ms)
      ✓ should return merged FormData if existing configMap is passed
      ✓ should return key value pairs for configmap with binaryData field
      ✓ should return key value pairs for configmap with both data and binaryData field (1ms)
    getConfigMapKeyValues
      ✓ should return empty array for invalid values 
      ✓ should return empty array for configmap without data or binaryData fields 
      ✓ should return key value pairs for configmap with data field
      ✓ should return key value pairs for configmap with binaryData field
    getConfigMapInitialValues:
      ✓ should return the default initial values (2ms)
      ✓ should return the initial values based on passed configmap (1ms)
    sanitizeToForm:
      ✓ should return default initial form data if configmap is not passed
      ✓ should convert the object from yaml to formData (1ms)
      ✓ should convert the binaryData value to readable format
    sanitizeToYaml:
      ✓ should return default yaml values for null input (1ms)
      ✓ should convert the object from yaml to formData
      ✓ should convert the binaryData value to base64 format when switched to yamlview (1ms)
      ✓ should contain original metadata labels and annotations if the configmap has it (1ms)
  ```
**Browser conformance**: 
<!-- To mark tested browsers, use [x] -->
- [x] Chrome
- [x] Firefox
- [x] Safari
- [ ] Edge

cc: @openshift/team-devconsole-ux @beaumorley @serenamarie125

karthikjeeyar commented:
/assign @debsmita1 @jerolimov

nemesis09 commented:
tested locally. 
- could create and edit a configmap succcessfully.
- edit is disabled with error message when immutable is set to true.
- honors user preference for editor

Noticed a css issue with the `Remove key/value` button. There is an extra spacing of `16px` due to the `.pf-l-flex.pf-m-column` class.
![Screenshot from 2022-03-08 15-08-10](https://user-images.githubusercontent.com/38663217/157214011-00a072d0-3f93-443d-9cd4-e7d649927e8a.png)

karthikjeeyar commented:
@nemesis09  I have updated the styles.
<img width="782" alt="image" src="https://user-images.githubusercontent.com/9964343/157257552-3e51150c-baf7-4fd8-8df5-f0b0a34bcecd.png">

serenamarie125 commented:
Functionality looks great!  I'd like to see a couple of UX improvements:
1 - text/help text for **immutable**
2 - is the key/value pair with the Remove link a common pattern?  If so, it's fine.  If not, I think we should improve it as it's a bit tough to know which key/value pair it's associated with.

@beaumorley FYI

serenamarie125 commented:
Wonder if we could do something like this, but have the tooltip be Remove key/value
![Screen Shot 2022-03-08 at 12 01 32 PM](https://user-images.githubusercontent.com/12504403/157287985-531311c3-f311-4a3a-9b07-2b10cad42954.png)

@karthikjeeyar should `value` have a required indicator?

karthikjeeyar commented:
> Functionality looks great! I'd like to see a couple of UX improvements:
> 1 - text/help text for immutable

Yes, will follow up and change it.

> 2 - is the key/value pair with the Remove link a common pattern? If so, it's fine. If not, I think we should improve it as it's > a bit tough to know which key/value pair it's associated with.

Today in Admin console, Secret Key value form has this pattern and since we need this configmap form in admin side as well, I have followed the same pattern here for consistency.

**Create Secret (Key value) Form:**
---
<img width="746" alt="image" src="https://user-images.githubusercontent.com/9964343/157369624-5033a129-b9a4-40aa-84c6-d58fe3b2c720.png">

---

> @karthikjeeyar should value have a required indicator?

No, value is an optional field.

<img width="893" alt="image" src="https://user-images.githubusercontent.com/9964343/157370693-08af8eb7-64d3-436b-8205-b8b5cf0ac410.png">
@serenamarie125 Let me know if you would prefer to stick with secret's form key value pattern or update the PR  to match above pattern.

cc: @beaumorley

invincibleJai commented on frontend/packages/console-shared/src/components/formik-fields/key-value-file-input-field/KeyValueFileInputField.tsx:
```diff
@@ -0,0 +1,103 @@
+  const { values } = useFormikContext<FormikValues>();
+  const rowValues = field.value ?? entries;
+  const fieldId = getFieldId(name, 'key-value--input');
+  const fieldValues = _.get(values, name, rowValues);
```
nit: can avoid lodash here

invincibleJai commented on frontend/public/components/configmaps/ConfigMapForm.tsx:
```diff
@@ -0,0 +1,93 @@
+import * as _ from 'lodash';
+import { Formik, FormikHelpers } from 'formik';
+import { useTranslation } from 'react-i18next';
+import { k8sCreate, k8sUpdate } from '@console/internal/module/k8s';
```
can use `k8sCreateResource` / `k8sUpdateResource` with new signature

karthikjeeyar replied:
updated

invincibleJai commented on frontend/public/components/configmaps/ConfigMapForm.tsx:
```diff
@@ -0,0 +1,93 @@
+import { Formik, FormikHelpers } from 'formik';
+import { useTranslation } from 'react-i18next';
+import { k8sCreate, k8sUpdate } from '@console/internal/module/k8s';
+import { ConfigMapModel } from '@console/internal/models';
+import { history } from '@console/internal/components/utils';
```
nit: can we do relative imports here as this in same package `public`

invincibleJai commented on frontend/public/components/configmaps/ConfigMapFormEditor.tsx:
```diff
@@ -0,0 +1,125 @@
+  SyncedEditorField,
+  YAMLEditorField,
+} from '@console/shared';
+
```
nit: new line not needed

invincibleJai commented on frontend/public/components/configmaps/ConfigMapForm.tsx:
```diff
@@ -0,0 +1,93 @@
+  ) => {
+    let resourceCall;
+    let configMapYaml: ConfigMap;
+    try {
+      // Use YAML also as base when submitting the form
+      configMapYaml = safeYAMLToJS(values.yamlData);
+    } catch (err) {
+      actions.setStatus({
+        submitSuccess: '',
+        submitError: t('public~Invalid YAML - {{err}}', { err }),
+      });
+      return;
+    }
```
do we need try / catch block as `safeYAMLToJS` has handled it and will return `{}` otherwise

karthikjeeyar replied:
No looks like we dont need try/catch as safeYAMLToJS already handles try/catch in its implementation. Removed it.

karthikjeeyar commented:
/retest

jerolimov commented on frontend/public/locales/en/public.json:
```diff
@@ -1710,7 +1719,7 @@
   "Self-support": "Self-support",
   "Self-support, 60 day trial": "Self-support, 60 day trial",
   "{{count}} day remaining": "{{count}} day remaining",
-  "{{count}} day remaining_plural": "{{count}} days remaining",
+  "{{count}} day remaining_plural": "{{count}} day remainings",
```
This change surprises me, can you check why this value is changed?

jerolimov commented on frontend/public/locales/en/public.json:
```diff
@@ -1719,7 +1728,7 @@
   "Your 60-day self-support trial expired.": "Your 60-day self-support trial expired.",
   "Your 60-day self-support trial will end in < 1 day.": "Your 60-day self-support trial will end in < 1 day.",
   "Your 60-day self-support trial will end in {{count}} day on {{date}}.": "Your 60-day self-support trial will end in {{count}} day on {{date}}.",
-  "Your 60-day self-support trial will end in {{count}} day on {{date}}._plural": "Your 60-day self-support trial will end in {{count}} days on {{date}}.",
+  "Your 60-day self-support trial will end in {{count}} day on {{date}}._plural": "Your 60-day self-support trial will end in {{count}} day on {{date}}.s",
```
This change surprises me, can you check why this value is changed?

karthikjeeyar replied:
Fixed it.

jerolimov commented on frontend/public/locales/en/public.json:
```diff
@@ -1719,7 +1728,7 @@
   "Your 60-day self-support trial expired.": "Your 60-day self-support trial expired.",
   "Your 60-day self-support trial will end in < 1 day.": "Your 60-day self-support trial will end in < 1 day.",
   "Your 60-day self-support trial will end in {{count}} day on {{date}}.": "Your 60-day self-support trial will end in {{count}} day on {{date}}.",
-  "Your 60-day self-support trial will end in {{count}} day on {{date}}._plural": "Your 60-day self-support trial will end in {{count}} days on {{date}}.",
+  "Your 60-day self-support trial will end in {{count}} day on {{date}}._plural": "Your 60-day self-support trial will end in {{count}} day on {{date}}.s",
```
This change surprises me, can you check why this value is changed?

jerolimov reviewed (changes requested):
@karthikjeeyar When I enter a string value (asdf: asdf) my value is saved as `binaryData` instead of `stringData`. Can you please take a look?

https://user-images.githubusercontent.com/139310/158157707-e9949744-a21f-48a7-bd68-b50a6dc8859d.mp4

There are also some unrelated i18n file changes, or?

karthikjeeyar commented:
@karthikjeeyar When I enter a string value (asdf: asdf) my value is saved as binaryData instead of stringData. Can you please take a look?

The problem is multiples of 4 is considered as a valid base64 string
eg:
```
Base64.encode('Aad')   => 'QWFk'
```
So `isBase64` string check (btoa(atob(str)) === str) returns true for strings that are multiples of 4. I will enquire in the channel if we can have multiple sections for data and binary data, so if we have explicit sections then it will straight forward instead of we detecting the user input for base64 implicitly.

nemesis09 commented:
/retest

karthikjeeyar commented:
/retest

debsmita1 commented on frontend/public/components/app-contents.tsx:
```diff
@@ -380,6 +380,26 @@ const AppContents: React.FC<{}> = () => {
           )
         }
       />
+      <LazyRoute
```
should we move the config map form components to dev-console package and use the dynamic extension for these routes

karthikjeeyar replied:
`configmap` is a standard K8s object like secrets, routes etc. We should keep it in public (console/internal) package.

debsmita1 replied:
I saw buildconfig form components are under dev-console package
@jerolimov ^

debsmita1 commented:
the Create Configmap button under all projects still shows the yaml editor

![cm](https://user-images.githubusercontent.com/22490998/160125348-d274c08f-cc1a-4f58-a86d-33240d54da94.gif)

karthikjeeyar commented:
@debsmita1 Ideally if we create any resource when All projects is selected in Administrator view, then we should redirect to default namespace. Thanks for pointing out. I have Updated it

https://user-images.githubusercontent.com/9964343/160168070-7a9425b7-1e58-4f30-ba84-ea2f16fa6c4c.mov

karthikjeeyar commented on frontend/public/components/app-contents.tsx:
```diff
@@ -380,6 +380,26 @@ const AppContents: React.FC<{}> = () => {
+        }
+      />
+      <LazyRoute
+        path="/k8s/ns/:ns/configmaps/:name/edit"
```
Updated

karthikjeeyar commented:
/retest

debsmita1 commented on frontend/packages/console-shared/src/components/formik-fields/key-value-file-input-field/KeyValueFileInputField.scss:
```diff
@@ -0,0 +1,9 @@
+}
+.key-value--wrapper + .key-value--wrapper {
+  border-top: 1px solid var(--pf-global--BorderColor--100);
+  padding-top: 10px;
```
can we use pf variable here ?

karthikjeeyar commented on frontend/public/components/app-contents.tsx:
```diff
@@ -380,6 +380,26 @@ const AppContents: React.FC<{}> = () => {
         }
       />
+      <LazyRoute
+        path="/k8s/ns/:ns/configmaps/~new/"
```
updated

debsmita1 commented:
there's an issue with the key-value pair
![Peek 2022-03-28 21-36](https://user-images.githubusercontent.com/22490998/160440488-7b608e12-e13d-4d24-bbf5-ff2a73ad795c.gif)

karthikjeeyar commented:
/retest

karthikjeeyar commented:
/retest

karthikjeeyar commented:
Fixed the issues with key/value pair and also update the dragAndDrop context to use single instance of HTML5 backend.

debsmita1 commented:
works fine now
but some tests are still failing

invincibleJai commented on frontend/public/locales/en/public.json:
```diff
@@ -281,6 +281,21 @@
+  "This object has been updated.": "This object has been updated.",
+  "Click reload to see the new version.": "Click reload to see the new version.",
+  "Save": "Save",
+  "A unique name for the configmap within the project": "A unique name for the configmap within the project",
```
we can be consistent, below I see `ConfigMap` and it make sense as we are referring to k8s resource
```suggestion
  "A unique name for the ConfigMap within the project": "A unique name for the ConfigMap within the project",
```

karthikjeeyar commented on frontend/public/components/configmaps/ConfigMapForm.tsx:
```diff
@@ -0,0 +1,89 @@
+    let resourceCall;
+
+    const configMapYaml: ConfigMap = safeYAMLToJS(values.yamlData);
+    if (configMapYaml?.metadata?.namespace) {
+      configMapYaml.metadata.namespace = namespace;
+    }
```
Updated the condition. The check should be to check if its not available

karthikjeeyar commented on frontend/public/components/nav/PinnedResource.tsx:
```diff
@@ -95,14 +96,14 @@ const PinnedResource: React.FC<PinnedResourceProps> = ({
       isOver: monitor.isOver(),
     }),
-    hover(item: DragItem) {
+    hover: debounce((item: DragItem) => {
```
Had to add debounce to avoid "Expected to find a valid target" error in react-dnd.

karthikjeeyar commented on frontend/public/components/configmaps/ConfigMapFormEditor.tsx:
```diff
@@ -0,0 +1,128 @@
+        infoTitle={t('public~This object has been updated.')}
+        infoMessage={t('public~Click reload to see the new version.')}
+        isSubmitting={isSubmitting}
+        submitLabel={t('public~Save')}
```
Updated to show `Create` on create flow and `Save` on edit flow.

karthikjeeyar commented:
Propagating docs/px approvals from the epic
/label docs-approved
/label px-approved

rottencandy reviewed (commented):
We should not be removing metadata from yaml when switching.

https://user-images.githubusercontent.com/20013884/161227789-b542bed1-9727-4dd4-8317-5216209c62f1.mp4

karthikjeeyar commented:
Fixed the edit flow to show all the metadata information from the existing configmap and added tests for that.
<img width="922" alt="image" src="https://user-images.githubusercontent.com/9964343/161240622-f7d710c2-ef0b-4cd2-9c77-70b1d3f8f33b.png">

rottencandy reviewed (approved):
/lgtm

Verified on cluster.
/label qe-approved

invincibleJai commented:
/approve

karthikjeeyar commented:
/retest
